src/main/scala/tech/cryptonomic/conseil/tezos/FeeOperations.scala

```diff
@@ -37,18 +45,20 @@ object FeeOperations extends LazyLogging {
     * Calculates average fees for each operation kind and stores them into a fees table.
     * @return
     */
-  def processTezosAverageFees(): Try[Unit] = {
+  def processTezosAverageFees()(implicit ex: ExecutionContext): Future[Option[Int]] = {
     logger.info("Processing latest Tezos fee data...")
-    val operationKinds = List("seed_nonce_revelation", "delegation", "transaction", "activate_account", "origination", "reveal", "double_endorsement_evidence", "endorsement")
-    val fees = operationKinds.map(TezosDatabaseOperations.calculateAverageFees)
-    Try {
-      val dbFut = TezosDatabaseOperations.writeFeesToDatabase(fees, db)
-      dbFut onComplete {
-        case Success(_) => logger.info(s"Wrote average fees to the database.")
-        case Failure(e) => logger.error(s"Could not write average fees to the database because $e")
+    val computeAndStore = for {
+      fees <- DBIOAction.sequence(operationKinds.map(Tdb.calculateAverageFeesIO))
+      dbWrites <- Tdb.writeFeesIO(fees.collect{ case Some(fee) => fee })
+    } yield dbWrites
```

Comment:
I'm guessing the idea here is to leave as many computations in our asynchronous computation in the for expression as possible? Besides good practice and readability, does this have performance benefits? (It definitely looks a lot nicer! :) )

Reply:
in this case Slick should optimize the calls, trying to offload the computation to the database when possible or in any case by using callbacks to actually make the second call only after the first result is available.

So yes, you're guessing right, slick works at 3 levels: Queries, DBActions and Future results and all of those can be composed as monads (using `flatMap`, `map`, `for`).

Additionally, composing `Query` objects before calling `.result` will deliver the composed query to the database, that will convert it to sql for "single-shot" execution, and for the database engine to optimize it

Reply:
Awesome! We'll have to talk about this type of refactor for APIOperations.scala. Lots of room for optimization there :)